pkg/controller/database/rdsinstance.go

```diff
@@ -92,47 +83,21 @@ func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.E
 
 	// TODO(negz): This connection logic should be generalised once this
 	// provider has more than one kind of managed resource.
-	var (
-		sel    *xpv1.SecretKeySelector
-		region string
-	)
-	switch {
-	case cr.GetProviderConfigReference() != nil:
-		if err := c.usage.Track(ctx, mg); err != nil {
-			return nil, errors.Wrap(err, errTrackUsage)
-		}
-
-		pc := &aliv1alpha1.ProviderConfig{}
-		if err := c.client.Get(ctx, types.NamespacedName{Name: cr.Spec.ProviderConfigReference.Name}, pc); err != nil {
-			return nil, errors.Wrap(err, errGetProviderConfig)
-		}
-		if s := pc.Spec.Credentials.Source; s != xpv1.CredentialsSourceSecret {
-			return nil, errors.Errorf(errFmtUnsupportedCredSource, s)
-		}
-		sel = pc.Spec.Credentials.SecretRef
-		region = pc.Spec.Region
-	case cr.GetProviderReference() != nil:
-		p := &aliv1alpha1.Provider{}
-		if err := c.client.Get(ctx, types.NamespacedName{Name: cr.Spec.ProviderReference.Name}, p); err != nil {
-			return nil, errors.Wrap(err, errGetProvider)
-		}
-		sel = p.Spec.CredentialsSecretRef
-		region = p.Spec.Region
-	default:
-		return nil, errors.New(errNoProvider)
+	if err := c.usage.Track(ctx, mg); err != nil {
+		return nil, errors.Wrap(err, errTrackUsage)
 	}
```

Comment:
You'll end up doing this (and most of the following steps) in all controllers so you may consider bundling them into a common implementation so you don't have to call `Track`, `GetCredentials`, `GetRegion` individually each time.

Reply:
Great solution. Done.